launch/launch/actions/group_action.py

```diff
@@ -48,12 +49,14 @@ def __init__(
         *,
         scoped: bool = True,
         launch_configurations: Optional[Dict[SomeSubstitutionsType, SomeSubstitutionsType]] = None,
+        forwarded_configurations: Optional[List[SomeSubstitutionsType]] = None,
```

Comment:
@djchopp hmm, I perceive some redundancy between `launch_configurations` and `forwarded_configurations`. IIUC, the only use case not covered by the current `GroupAction` is a non-forwarding, scoped group.

Can we achieve the same using a `forwarding` boolean, set to `True` by default?

Reply:
My understanding of the current `GroupAction` is that it does not support the "scoped and non-forwarding" or "scoped and partial-forwarding" cases. I think it is possible to merge `forwarded_configurations` into `launch_configurations` and add add the boolean for `forwarding` while supporting "scoped and non-forwarding", "scoped and partial-forwarding", "scoped and full-forwarded", and "unscoped" using the following table:

| Value of `scoped` | Value of `forwarded` | Key/Value pair in `launch_configurations` | Result |
| --- | --- | --- | --- |
| `True` | `True` or `False` | `Arg1: 'NewValue'` | `Arg1` will have value `NewValue` in the scoped context regardless if it existed or not |
| `True` | `True` | `Arg1: None` | `Arg1` will be passed through to the scoped context with its existing value |
| `True` | `False` | `Arg1: None` | No action is taken, but the end result is `Arg1` still being passed through to the scoped context |
| `True` | `True` | Key for an existing launch configuration does not exist in the `launch_configurations` dict | The key/value will be cleared from the scoped context |
| `True` | `False` | Key for an existing launch configuration does not exist in the `launch_configurations` dict | The key/value will still pass through to the scoped context unmodified |
| `False` | Does not matter |  | The group actions are not scoped so no context launch configurations are modified |

Using `Arg1: None` to signify forwarding may be a little confusing, but I am not sure how else to support partial-forwarding without it. Open to idea's. In any case I can provide a docstring to help describe the behavior

Reply:
@djchopp I'd suggest a simpler approach. Groups (and includes) can be non-scoped, scoped and non-forwarding, and scoped and forwarding. "Partial-forwarding" (`'Arg1': None` here) and explicit argument passing (`'Arg1': LaunchConfiguration('Arg1')`) are functionally equivalent and the latter reads better, which is why I think we do not need the former nor two lists.

Reply:
I agree the later is more intuitive, but unfortunately when forwarding is `False` the context would already be pushed and cleared by the time the resulting `SetLaunchConfiguration` from `launch_configurations={'Arg1': LaunchConfiguration('Arg1')}` is evaluated. The resulting list of actions would be:

    # scoped=True pushes the configurations at the start of the group action
    PushLaunchConfigurations(),

    # forwarding=False clears 'Arg1' from the launch configurations
    ClearLaunchConfigurations(), 

    # launch_configurations={'Arg1': LaunchConfiguration('Arg1')} will
    # add this SetLaunchConfiguration action to the list of actions,
    # but LaunchConfiguration('Arg1') cannot be resolved as it no longer exists
    SetLaunchConfiguration('Arg1', LaunchConfiguration('Arg1')), 

    # scoped=True pops the configurations at the end of the group action
    PopLaunchConfigurations()

Therefore ending in an error: `launch.substitutions.substitution_failure.SubstitutionFailure: launch configuration 'Arg1' does not exist`. 

One possibility around this is to evaluate the substitutions before clearing the context. Essentially evaluate the substitutions into strings that way they are not tied to a `LaunchConfiguration` object by the time the `SetLaunchConfiguration` is executed. Do you foresee any issue with this approach?

The issue I see is that I don't think this pattern would be possible outside of the group action since it requires evaluation of substitutions in a non-standard order compared to how the base actions (`Push`, `Clear`, `Set`, `Pop`) operate.

Reply:
>  but unfortunately when forwarding is False the context would already be pushed and cleared by the time the resulting SetLaunchConfiguration from launch_configurations={'Arg1': LaunchConfiguration('Arg1')} is evaluated.

You're absolutely right. There's dragons in handling scope as a global dictionary.

> The issue I see is that I don't think this pattern would be possible outside of the group action since it requires evaluation of substitutions in a non-standard order compared to how the base actions (Push, Clear, Set, Pop) operate.

Hmm, that's also true. What about collapsing the `ClearLaunchConfigurations` and `SetLaunchConfiguration` actions into a single `ResetLaunchConfigurations` action that can force a given context? We could bridge "outer" and "inner" contexts in there, and `ClearLaunchConfigurations` would become `ResetLaunchConfigurations` with no arguments. Your draft of a partial-forwarding scoped group implementation would then be:

```
# scoped=True pushes the configurations at the start of the group action
PushLaunchConfigurations(),

# forwarding=False resets the context to match the given 
# launch_configurations={'Arg1': LaunchConfiguration('Arg1')} will.
# If no launch_configurations are given, the clear-all effect is achieved
ResetLaunchConfigurations(launch_configurations),

# scoped=True pops the configurations at the end of the group action
PopLaunchConfigurations()
```

I'm not amazed by the idea but the gap seems small enough. WDYT @djchopp?

Reply:
@hidmic I like it better than my other idea which was to allow access to launch configurations up the stack with a new `OutOfScopeLaunchConfiguration` substitution type. For the record it did work and would look like  `launch_configurations={'Arg1': OutOfScopeLaunchConfiguration('Arg1', depth=1)}`, but it just feels pretty far in the weeds / playing with fire. I will work on a `ResetLaunchConfigurations` action and see if I find any issues.